File: backend/lumen_schemas/knowledge.py
from pydantic import BaseModel, Field
from datetime import datetime
from typing import Optional, List, Dict, Literal


class KnowledgeBaseBase(BaseModel):
    name: str
    description: Optional[str] = None
    # Deprecated: stored on the row for backwards compatibility, but the
    # backend now drives embedding model from `embedding_model_config_id`.
    embedding_model: Optional[str] = Field(
        None,
        description="DEPRECATED: use embedding_model_config_id. Read-only fallback.",
    )
    embedding_model_config_id: Optional[int] = Field(
        None,
        description="Embedding ModelConfig id. Locked once the KB is created.",
    )
    search_weights: Optional[Dict[str, float]] = None  # e.g. {"title": 10.0, "text": 2.0}
    default_parser: Optional[str] = "general"
    chunk_size: int = 500
    chunk_overlap: int = 50
    # M38.4: 多模态开关。开启后,KB 内的 image 文档会同时写入独立的
    # multimodal FAISS 索引(`data/multimodal/kb-{id}-mm.faiss`),并
    # 支持 `POST /image-search`(image-to-image 检索)。
    multimodal_enabled: bool = Field(
        False,
        description="M38.4 多模态开关: True 后 KB 内的图片 chunk 会走独立 multimodal 向量库",
    )
    multimodal_config_id: Optional[int] = Field(
        None,
        description="M38.4 绑定的 multimodal config id,multimodal_enabled=True 时必填",
    )


class KnowledgeBaseCreate(KnowledgeBaseBase):
    embedding_model_config_id: int = Field(
        ...,
        description="Required: FK to ModelConfig.id. Locked after creation.",
    )


class KnowledgeBaseUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    # NOTE: embedding_model_config_id is intentionally absent. The PUT
    # endpoint also rejects changes defensively in case a future
    # schema drift reintroduces the field.
    search_weights: Optional[Dict[str, float]] = None
    default_parser: Optional[str] = None
    chunk_size: Optional[int] = None
    chunk_overlap: Optional[int] = None
    # M38.4: 多模态开关与绑定的 config id。两者互不影响 — 用户可
    # 单开 multimodal_enabled 后再绑 config(开但没 config 时新上传
    # image 文档不会被 multimodal 索引,gracefully degraded)。
    multimodal_enabled: Optional[bool] = None
    multimodal_config_id: Optional[int] = None


class KnowledgeBaseResponse(KnowledgeBaseBase):
    id: int
    tenant_id: int
    status: str
    created_at: datetime
    default_parser: Optional[str] = "general"
    chunk_size: int = 500
    chunk_overlap: int = 50
    # Number of documents in this KB. Populated by the service layer
    # via a single GROUP BY query — kept as a transient attribute on the
    # SQLAlchemy instance so Pydantic reads it via from_attributes.
    document_count: int = 0

    class Config:
        from_attributes = True


class DocumentResponse(BaseModel):
    id: int
    filename: str
    file_type: str
    file_size: int
    status: str
    chunk_count: Optional[int] = None
    error_message: Optional[str] = None
    created_at: datetime
    knowledge_base_id: Optional[int] = None

    class Config:
        from_attributes = True


class ChunkResponse(BaseModel):
    id: int
    content: str
    chunk_index: int
    vector_id: Optional[str] = None
    # Surfaced to the UI so it can show "10/14 chunks indexed" or
    # grey out rows that the embedder rejected. Frontend uses
    # string equality with the backend (currently 'ok' | 'failed').
    embedding_status: Optional[str] = None
    # M38.4: 模态字段 —— text / image / audio / video。默认 'text'
    # 让 legacy chunk 行为不变;Excel 走 'text' + sheet_name;PPT
    # 抽出图走 'image' + page_number + image_caption。frontend 用
    # modality 决定要不要走 image gallery 而不是 text diff view。
    modality: str = "text"
    sheet_name: Optional[str] = None
    page_number: Optional[int] = None
    image_caption: Optional[str] = None

    class Config:
        from_attributes = True


class RechunkRequest(BaseModel):
    """Body for POST /knowledge/documents/{id}/rechunk.

    All fields are optional — unset fields fall back to the doc's
    currently stored values (or the parent KB's chunk_size/overlap
    if the doc doesn't store them).
    """
    chunking_strategy: Optional[str] = "fixed"  # fixed | semantic | document_structure
    chunk_size: Optional[int] = None
    chunk_overlap: Optional[int] = None
    doc_type: Optional[str] = None  # general | paper | qa | table | manual | laws


# ------------------------------------------------------------------ M31: FAQ Q&A

class FAQEntryBase(BaseModel):
    """Common shape shared by FAQEntryCreate / FAQEntryUpdate.

    Field-length limits are chosen to match the underlying
    schema (question/answer stored as TEXT — soft caps protect
    against accidental 10MB inputs from the UI textareas).
    """
    question: str = Field(..., min_length=1, max_length=2000)
    answer: str = Field(..., min_length=1, max_length=8000)
    # Free-form category like "退货政策" / "物流时效". Stored
    # as VARCHAR(50) in the FAQEntry table.
    category: Optional[str] = Field(None, max_length=50)
    # Free-form tags list (e.g. ["急", "VIP"]). Not bounded by
    # a fixed max_items here; FAQService.create_entry enforces
    # < 20 tags to match the bulk-import validator.
    tags: Optional[List[str]] = None


class FAQEntryCreate(FAQEntryBase):
    """Body for POST /knowledge/{kb_id}/faq-entries (single create)."""
    pass


class FAQEntryUpdate(BaseModel):
    """Body for PUT /knowledge/{kb_id}/faq-entries/{id}.

    Every field is optional — the UI sends a PATCH-style update
    with only the changed fields. The service layer treats None
    as "leave unchanged" and only updates fields that are
    present in the payload.
    """
    question: Optional[str] = Field(None, min_length=1, max_length=2000)
    answer: Optional[str] = Field(None, min_length=1, max_length=8000)
    category: Optional[str] = Field(None, max_length=50)
    tags: Optional[List[str]] = None


class FAQEntryResponse(FAQEntryBase):
    """Response shape for a single FAQ row.

    Mirrors the FAQEntry ORM model 1:1 — vector_id, document_id,
    chunk_id are exposed so the UI can navigate to the related
    resources if needed (e.g. for debugging chunk-level issues).
    """
    id: int
    knowledge_base_id: int
    vector_id: Optional[str] = None
    document_id: int
    chunk_id: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class FAQBulkImportRequest(BaseModel):
    """Body for POST /knowledge/{kb_id}/faq-entries/bulk.

    Two formats are supported:
    - ``"json"``: a JSON array of objects with the same shape
      as FAQEntryCreate (question, answer, category?, tags?).
    - ``"csv"``: CSV text with header row
      ``question,answer,category,tags`` (tags is a comma-separated
      string within the cell).
    """
    format: Literal["json", "csv"]
    content: str = Field(
        ...,
        description="JSON array or CSV text. For CSV use header row: question,answer,category,tags.",
    )


class FAQBulkImportResult(BaseModel):
    """Result of a bulk import call.

    Two failure surfaces:
    - ``failed`` collects per-row validation errors (empty
      question, malformed JSON, etc.). These are non-fatal — the
      valid rows still get inserted.
    - Anything that fails at the DB / vector-store layer raises
      a 500 and the whole transaction is rolled back. We don't
      try to "best-effort" partial inserts because the failure
      mode is typically "embedder down" and a half-written
      transaction is worse than an all-or-nothing one.
    """
    inserted: int
    failed: List[Dict[str, str]] = Field(
        default_factory=list,
        description="Per-row validation errors: [{row_index, reason}, ...]",
    )


# -------------------------------------------------------------- M38.4: Images


class ImageAssetResponse(BaseModel):
    """One image row for the gallery endpoint.

    镜像 ``ImageAsset`` ORM 1:1 + ``doc_filename`` 通过 JOIN Document
    填进来,前端 gallery 直接渲染。``embedding_status`` 让 UI 知道
    「暂未索引」(multimodal embedder 失败时设 'failed')。
    """
    id: int
    document_id: int
    chunk_id: Optional[int] = None
    # PPT 抽出图时存源页码(1-based);独立上传的图片为 NULL。
    original_doc_page: Optional[int] = None
    storage_key: str
    width: Optional[int] = None
    height: Optional[int] = None
    mime_type: Optional[str] = None
    file_size: Optional[int] = None
    caption: Optional[str] = None
    embedding_status: str = "pending"  # pending | ok | failed
    doc_filename: Optional[str] = None
    created_at: datetime

    class Config:
        from_attributes = True


class ImageSearchHit(BaseModel):
    """One hit returned by ``POST /knowledge/{kb_id}/image-search``."""
    asset_id: Optional[int] = None
    chunk_id: Optional[int] = None
    document_id: Optional[int] = None
    doc_filename: Optional[str] = None
    image_caption: Optional[str] = None
    storage_key: Optional[str] = None
    # 内积距离(FAISS IndexFlatIP)。score = -distance,值越大越相似。
    distance: Optional[float] = None
    score: Optional[float] = None


class ImageSearchResponse(BaseModel):
    """Body of ``POST /knowledge/{kb_id}/image-search``.

    ``search_mode`` 区分走的是 multimodal embedding(真 image-to-image
    检索)还是 text fallback —— fallback 时 admin UI 应提示用户「多模态
    模型未就绪,已用文字描述匹配」,但不抛 503(graceful degradation)。
    """
    search_mode: str  # "multimodal" | "text_fallback"
    query_caption: Optional[str] = None
    results: List[ImageSearchHit] = Field(default_factory=list)


class ImageUploadResponse(BaseModel):
    """Body of ``POST /knowledge/{kb_id}/images/upload``."""
    document_id: int
    task_id: Optional[str] = None
    status: str

File: frontend/services/knowledge.ts
import api from "./auth";
import type { KnowledgeBase, ApiResponse, PaginatedResponse } from "@/types/api";

export interface ParserType {
  type: string;
  label: string;
  description: string;
}

export interface SearchOptions {
  k?: number;
  alpha?: number;
  rerank?: boolean;
  rerank_top_n?: number;
  field_weights?: string;
}

export interface KnowledgeBaseUpdate {
  name?: string;
  description?: string;
  // embedding_model is a legacy string field (e.g. "nomic-embed-text").
  // New code should rely on embedding_model_config_id on the response
  // shape, not send this on update. The PUT endpoint ignores it.
  embedding_model?: string;
  // Note: embedding_model_config_id is intentionally NOT in the
  // update payload — the embedding model is locked once a KB is
  // created. Sending it has no effect (the API strips it).
  search_weights?: Record<string, number>;
  default_parser?: string;
  chunk_size?: number;
  chunk_overlap?: number;
}

export interface KnowledgeBaseCreate {
  name: string;
  description?: string;
  // The ModelConfig.id of the embedding model to use. Required on
  // create; the API rejects requests without it with a 422.
  embedding_model_config_id: number;
  // Legacy string kept for backward compat with old API responses
  // (read-only on the response shape). The frontend form no longer
  // collects this.
  embedding_model?: string;
  search_weights?: Record<string, number>;
  default_parser?: string;
  chunk_size?: number;
  chunk_overlap?: number;
}

export interface DocumentResponse {
  id: number;
  filename: string;
  file_type: string;
  file_size: number;
  status: string;
  chunk_count?: number;
  error_message?: string;
  created_at: string;
  knowledge_base_id?: number;
  doc_metadata?: {
    doc_type?: string;
  };
}

export interface DocumentChunk {
  id: number;
  chunk_index: number;
  content: string;
  vector_id?: string;
}

export interface RechunkParams {
  chunking_strategy?: string;
  chunk_size?: number;
  chunk_overlap?: number;
  doc_type?: string;
}

// ----------------------------------------------------------------- M31: FAQ Q&A
//
// The Q&A tab on the KB detail page is driven by these
// types. The API is 1:1 with the backend FAQService
// (see backend/app/services/faq_service.py).

export interface FAQEntry {
  id: number;
  knowledge_base_id: number;
  question: string;
  answer: string;
  category?: string;
  tags?: string[];
  vector_id?: string;
  document_id: number;
  chunk_id: number;
  created_at: string;
  updated_at: string;
}

export interface FAQEntryCreate {
  question: string;
  answer: string;
  category?: string;
  tags?: string[];
}

export interface FAQEntryUpdate {
  question?: string;
  answer?: string;
  category?: string;
  tags?: string[];
}

export interface FAQBulkImportRequest {
  format: "json" | "csv";
  content: string;
}

export interface FAQBulkImportResult {
  inserted: number;
  failed: { row_index: string; reason: string }[];
}

export const knowledgeApi = {
  list: (page = 1, pageSize = 10, workspaceId?: number) => {
    let url = `/knowledge/?page=${page}&page_size=${pageSize}`;
    // M38.2: workspace filter. ``workspaceId === 0`` means "tenant root
    // (no workspace)" — backend uses ``0`` as the sentinel.
    if (workspaceId !== undefined && workspaceId !== -1) {
      url += `&workspace_id=${workspaceId}`;
    }
    return api.get<PaginatedResponse<KnowledgeBase>>(url);
  },
  get: (id: number) => api.get<ApiResponse<KnowledgeBase>>(`/knowledge/${id}`),
  getDocuments: (kbId: number, folderId?: number) => {
    // M38.2: ``folder_id`` filter — 0 = KB root, -1/None = all, >0 = that folder.
    let url = `/knowledge/${kbId}/documents`;
    if (folderId !== undefined) {
      url += `?folder_id=${folderId}`;
    }
    return api.get<ApiResponse<DocumentResponse[]>>(url);
  },
  create: (data: KnowledgeBaseCreate, workspaceId?: number) => {
    // M38.2: optional workspace binding at create.
    let url = "/knowledge/";
    if (workspaceId !== undefined && workspaceId !== null) {
      url += `?workspace_id=${workspaceId}`;
    }
    return api.post<ApiResponse<KnowledgeBase>>(url, data);
  },
  update: (id: number, data: KnowledgeBaseUpdate) =>
    api.put<ApiResponse<KnowledgeBase>>(`/knowledge/${id}`, data),
  delete: (id: number) => api.delete(`/knowledge/${id}`),
  deleteDocument: (docId: number) =>
    api.delete<
      ApiResponse<{
        document_id: number;
        deleted_chunks: number;
        vector_cleanup_failed: boolean;
      }>
    >(`/knowledge/documents/${docId}`),
  upload: (kbId: number, file: File, docType?: string, folderId?: number) => {
    const formData = new FormData();
    formData.append("file", file);
    if (docType) {
      formData.append("doc_type", docType);
    }
    // M38.2: optional folder_id for upload.
    if (folderId !== undefined && folderId !== null) {
      formData.append("folder_id", String(folderId));
    }
    return api.post(`/knowledge/${kbId}/documents`, formData, {
      headers: { "Content-Type": "multipart/form-data" },
    });
  },
  search: (kbId: number, query: string, options?: SearchOptions) => {
    let url = `/knowledge/${kbId}/search?query=${encodeURIComponent(query)}`;
    if (options) {
      if (options.k !== undefined) url += `&k=${options.k}`;
      if (options.alpha !== undefined) url += `&alpha=${options.alpha}`;
      if (options.rerank !== undefined) url += `&rerank=${options.rerank}`;
      if (options.rerank_top_n !== undefined) url += `&rerank_top_n=${options.rerank_top_n}`;
      if (options.field_weights) url += `&field_weights=${encodeURIComponent(options.field_weights)}`;
    }
    return api.get<ApiResponse<any[]>>(url);
  },
  getParserTypes: () =>
    api.get<ApiResponse<{ parser_types: ParserType[]; chunking_strategies: any[] }>>("/knowledge/parser-types"),
  getDocumentStatus: (docId: number) =>
    api.get<ApiResponse<{ document_id: number; status: string; chunk_count?: number; error_message?: string }>>(`/knowledge/documents/${docId}/status`),
  retry: (docId: number) =>
    api.post<ApiResponse<DocumentResponse>>(`/knowledge/documents/${docId}/retry`),
  listChunks: (docId: number, page = 1, pageSize = 50) =>
    api.get<ApiResponse<DocumentChunk[]>>(
      `/knowledge/documents/${docId}/chunks?page=${page}&page_size=${pageSize}`
    ),
  rechunk: (docId: number, params: RechunkParams) =>
    api.post<ApiResponse<DocumentResponse>>(
      `/knowledge/documents/${docId}/rechunk`,
      params
    ),
  // M31: FAQ Q&A endpoints. The list endpoint returns a
  // PaginatedResponse (with the standard envelope), so the
  // generic is `PaginatedResponse<FAQEntry>`. The single
  // create / update / delete endpoints use the standard
  // SingleResponse envelope.
  listFaqs: (
    kbId: number,
    params?: {
      page?: number;
      page_size?: number;
      category?: string;
      search?: string;
    }
  ) => {
    const usp = new URLSearchParams();
    if (params?.page) usp.set("page", String(params.page));
    if (params?.page_size) usp.set("page_size", String(params.page_size));
    if (params?.category) usp.set("category", params.category);
    if (params?.search) usp.set("search", params.search);
    const qs = usp.toString();
    return api.get<PaginatedResponse<FAQEntry>>(
      `/knowledge/${kbId}/faq-entries${qs ? `?${qs}` : ""}`
    );
  },
  createFaq: (kbId: number, data: FAQEntryCreate) =>
    api.post<ApiResponse<FAQEntry>>(
      `/knowledge/${kbId}/faq-entries`,
      data
    ),
  updateFaq: (kbId: number, id: number, data: FAQEntryUpdate) =>
    api.put<ApiResponse<FAQEntry>>(
      `/knowledge/${kbId}/faq-entries/${id}`,
      data
    ),
  deleteFaq: (kbId: number, id: number) =>
    api.delete<ApiResponse<{ entry_id: number; deleted: boolean }>>(
      `/knowledge/${kbId}/faq-entries/${id}`
    ),
  bulkImportFaqs: (kbId: number, data: FAQBulkImportRequest) =>
    api.post<ApiResponse<FAQBulkImportResult>>(
      `/knowledge/${kbId}/faq-entries/bulk`,
      data
    ),
};
